File: typr-tester-typr-scala-new/src/scala/adventureworks/SeekDbTest.scala
package adventureworks

import adventureworks.person.businessentity.*
import org.junit.Assert.*
import org.junit.Test
import typr.runtime.PgTypes

import java.time.LocalDateTime
import java.util.UUID

class SeekDbTest {
  private val timestampPgType = PgTypes.timestamp.withTypename("timestamp")

  private def testUniformSeek(businessentityRepo: BusinessentityRepo): Unit = {
    val limit = 3
    val now = LocalDateTime.of(2021, 1, 1, 0, 0)

    val rows = (0 until limit * 2).map { i =>
      val time = now.minusDays(i % limit)
      BusinessentityRow(BusinessentityId(i), UUID.randomUUID(), time)
    }.toList

    val sortedRows = rows.sortBy(r => (r.modifieddate, r.businessentityid.value))
    val group1 = sortedRows.take(limit)
    val group2 = sortedRows.slice(limit, limit * 2)

    WithConnection {
      val _ = businessentityRepo.insertStreaming(rows.iterator, 100)

      val rows1 = businessentityRepo.select
        .maybeSeek(f => f.modifieddate.asc, Option.empty[LocalDateTime], (v: LocalDateTime) => new typr.dsl.SqlExpr.ConstReq(v, timestampPgType))
        .maybeSeek(f => f.businessentityid.asc, Option.empty[BusinessentityId], (v: BusinessentityId) => new typr.dsl.SqlExpr.ConstReq(v, BusinessentityId.pgType))
        .maybeSeek(f => f.rowguid.asc, Option.empty[UUID], (v: UUID) => new typr.dsl.SqlExpr.ConstReq(v, PgTypes.uuid))
        .limit(limit)
        .toList

      assertEquals(group1, rows1)

      val lastRow = rows1.last
      val rows2 = businessentityRepo.select
        .maybeSeek(f => f.modifieddate.asc, Some(lastRow.modifieddate), (v: LocalDateTime) => new typr.dsl.SqlExpr.ConstReq(v, timestampPgType))
        .maybeSeek(f => f.businessentityid.asc, Some(lastRow.businessentityid), (v: BusinessentityId) => new typr.dsl.SqlExpr.ConstReq(v, BusinessentityId.pgType))
        .maybeSeek(f => f.rowguid.asc, Some(lastRow.rowguid), (v: UUID) => new typr.dsl.SqlExpr.ConstReq(v, PgTypes.uuid))
        .limit(limit)
        .toList

      assertEquals(group2, rows2)
    }
  }

  @Test
  def uniformInMemory(): Unit = {
    testUniformSeek(BusinessentityRepoMock(_ => throw new UnsupportedOperationException))
  }

  @Test
  def uniformPg(): Unit = {
    testUniformSeek(BusinessentityRepoImpl())
  }

  private def testNonUniformSeek(businessentityRepo: BusinessentityRepo): Unit = {
    val limit = 3
    val now = LocalDateTime.of(2021, 1, 1, 0, 0)

    val rows = (0 until limit * 2).map { i =>
      val time = now.minusDays(i % limit)
      BusinessentityRow(BusinessentityId(i), UUID.randomUUID(), time)
    }.toList

    val sortedRows = rows.sortBy(r => (-r.modifieddate.toEpochSecond(java.time.ZoneOffset.UTC), r.businessentityid.value))
    val group1 = sortedRows.take(limit)
    val group2 = sortedRows.slice(limit, limit * 2)

    WithConnection {
      val _ = businessentityRepo.insertStreaming(rows.iterator, 100)

      val rows1 = businessentityRepo.select
        .maybeSeek(f => f.modifieddate.desc, Option.empty[LocalDateTime], (v: LocalDateTime) => new typr.dsl.SqlExpr.ConstReq(v, timestampPgType))
        .maybeSeek(f => f.businessentityid.asc, Option.empty[BusinessentityId], (v: BusinessentityId) => new typr.dsl.SqlExpr.ConstReq(v, BusinessentityId.pgType))
        .maybeSeek(f => f.rowguid.asc, Option.empty[UUID], (v: UUID) => new typr.dsl.SqlExpr.ConstReq(v, PgTypes.uuid))
        .limit(limit)
        .toList

      assertEquals(group1, rows1)

      val lastRow = rows1.last
      val rows2 = businessentityRepo.select
        .maybeSeek(f => f.modifieddate.desc, Some(lastRow.modifieddate), (v: LocalDateTime) => new typr.dsl.SqlExpr.ConstReq(v, timestampPgType))
        .maybeSeek(f => f.businessentityid.asc, Some(lastRow.businessentityid), (v: BusinessentityId) => new typr.dsl.SqlExpr.ConstReq(v, BusinessentityId.pgType))
        .maybeSeek(f => f.rowguid.asc, Some(lastRow.rowguid), (v: UUID) => new typr.dsl.SqlExpr.ConstReq(v, PgTypes.uuid))
        .limit(limit)
        .toList

      assertEquals(group2, rows2)
    }
  }

  @Test
  def nonUniformInMemory(): Unit = {
    testNonUniformSeek(BusinessentityRepoMock(_ => throw new UnsupportedOperationException))
  }

  @Test
  def nonUniformPg(): Unit = {
    testNonUniformSeek(BusinessentityRepoImpl())
  }
}

File: typr-tester-typr-kotlin/src/test/kotlin/adventureworks/SeekDbTest.kt
package adventureworks

import adventureworks.DbNow
import java.util.UUID
import adventureworks.person.businessentity.*
import org.junit.Assert.assertEquals
import org.junit.Test
import java.time.LocalDateTime

class SeekDbTest {
    private fun testUniformSeek(businessentityRepo: BusinessentityRepo) {
        val limit = 3
        val now = LocalDateTime.of(2021, 1, 1, 0, 0)
        val rows = (0 until limit * 2).map { i ->
            // ensure some duplicate values
            val time = now.minusDays((i % limit).toLong())
            BusinessentityRow(BusinessentityId(i), UUID.randomUUID(), time)
        }

        // same as sql below
        val sorted = rows.sortedWith(compareBy({ it.modifieddate }, { it.businessentityid.value }))
        val group1 = sorted.take(limit)
        val group2 = sorted.drop(limit).take(limit)

        WithConnection.run { c ->
            // batch insert some rows
            businessentityRepo.insertStreaming(rows.iterator(), 1000, c)
            val rows1 = businessentityRepo.select()
                .maybeSeek({ it.modifieddate().asc() }, null)
                .maybeSeek({ it.businessentityid().asc() }, null)
                .maybeSeek({ it.rowguid().asc() }, null)
                .limit(limit)
                .toList(c)
            assertEquals(group1, rows1)
            val rows2 = businessentityRepo.select()
                .maybeSeek({ it.modifieddate().asc() }, rows1.last().modifieddate)
                .maybeSeek({ it.businessentityid().asc() }, rows1.last().businessentityid)
                .maybeSeek({ it.rowguid().asc() }, rows1.last().rowguid)
                .limit(limit)
                .toList(c)
            assertEquals(group2, rows2)
        }
    }

    @Test
    fun uniformInMemory() {
        testUniformSeek(BusinessentityRepoMock(toRow = { it.toRow(
            businessentityidDefault = { BusinessentityId(0) },
            rowguidDefault = { UUID.randomUUID() },
            modifieddateDefault = { DbNow.localDateTime() }
        ) }))
    }

    @Test
    fun uniformPg() {
        testUniformSeek(BusinessentityRepoImpl())
    }

    private fun testNonUniformSeek(businessentityRepo: BusinessentityRepo) {
        val limit = 3
        val now = LocalDateTime.of(2021, 1, 1, 0, 0)
        val rows = (0 until limit * 2).map { i ->
            // ensure some duplicate values
            val time = now.minusDays((i % limit).toLong())
            BusinessentityRow(BusinessentityId(i), UUID.randomUUID(), time)
        }

        // same as sql below
        val sorted = rows.sortedWith(compareBy({ -it.modifieddate.toEpochSecond(java.time.ZoneOffset.UTC) }, { it.businessentityid.value }))
        val group1 = sorted.take(limit)
        val group2 = sorted.drop(limit).take(limit)

        WithConnection.run { c ->
            // batch insert some rows
            businessentityRepo.insertStreaming(rows.iterator(), 1000, c)
            val rows1 = businessentityRepo.select()
                .maybeSeek({ it.modifieddate().desc() }, null)
                .maybeSeek({ it.businessentityid().asc() }, null)
                .maybeSeek({ it.rowguid().asc() }, null)
                .limit(limit)
                .toList(c)
            assertEquals(group1, rows1)
            val rows2 = businessentityRepo.select()
                .maybeSeek({ it.modifieddate().desc() }, rows1.last().modifieddate)
                .maybeSeek({ it.businessentityid().asc() }, rows1.last().businessentityid)
                .maybeSeek({ it.rowguid().asc() }, rows1.last().rowguid)
                .limit(limit)
                .toList(c)
            assertEquals(group2, rows2)
        }
    }

    @Test
    fun nonUniformInMemory() {
        testNonUniformSeek(BusinessentityRepoMock(toRow = { it.toRow(
            businessentityidDefault = { BusinessentityId(0) },
            rowguidDefault = { UUID.randomUUID() },
            modifieddateDefault = { DbNow.localDateTime() }
        ) }))
    }

    @Test
    fun nonUniformPg() {
        testNonUniformSeek(BusinessentityRepoImpl())
    }
}
